packages/vite-plugin/src/index.ts

```diff
@@ -1,62 +1,98 @@
-  `vanilla-extract-style-update:${fileId}`;
 
 const virtualExtCss = '.vanilla.css';
-const virtualExtJs = '.vanilla.js';
+
+const isVirtualId = (id: string) => id.endsWith(virtualExtCss);
+const fileIdToVirtualId = (id: string) => `${id}${virtualExtCss}`;
+const virtualIdToFileId = (virtualId: string) =>
+  virtualId.slice(0, -virtualExtCss.length);
 
 interface Options {
   identifiers?: IdentifierOption;
-  emitCssInSsr?: boolean;
-  esbuildOptions?: CompileOptions['esbuildOptions'];
+  unstable_mode?: 'transform' | 'emitCss';
 }
 export function vanillaExtractPlugin({
   identifiers,
-  emitCssInSsr,
-  esbuildOptions,
+  unstable_mode: mode = 'emitCss',
 }: Options = {}): Plugin {
   let config: ResolvedConfig;
   let server: ViteDevServer;
-  let postCssConfig: PostCSSConfigResult | null;
-  // We lazily load this utility from Vite
-  let normalizePath: (fsPath: string) => string;
-  const cssMap = new Map<string, string>();
-
-  const hasEmitCssOverride = typeof emitCssInSsr === 'boolean';
-  let resolvedEmitCssInSsr: boolean = hasEmitCssOverride
-    ? emitCssInSsr
-    : !!process.env.VITE_RSC_BUILD;
   let packageName: string;
+  let compiler: Compiler | undefined;
+
+  const getIdentOption = () =>
+    identifiers ?? (config.mode === 'production' ? 'short' : 'debug');
+  const getAbsoluteId = (filePath: string) => {
+    let resolvedId = filePath;
+
+    if (
+      filePath.startsWith(config.root) ||
+      // In monorepos the absolute path will be outside of config.root, so we check that they have the same root on the file system
+      (path.isAbsolute(filePath) &&
+        filePath.split(path.sep)[1] === config.root.split(path.sep)[1])
+    ) {
+      resolvedId = filePath;
+    } else {
+      // In SSR mode we can have paths like /app/styles.css.ts
+      resolvedId = path.join(config.root, filePath);
+    }
+
+    return normalizePath(resolvedId);
+  };
 
-  const getAbsoluteVirtualFileId = (source: string) =>
-    normalizePath(path.join(config.root, source));
+  function invalidateModule(absoluteId: string) {
+    if (!server) return;
+
+    const { moduleGraph } = server;
+    const modules = Array.from(moduleGraph.getModulesByFile(absoluteId) || []);
+
+    for (const module of modules) {
+      moduleGraph.invalidateModule(module);
+
+      // Vite uses this timestamp to add `?t=` query string automatically for HMR.
+      module.lastHMRTimestamp = module.lastInvalidationTimestamp || Date.now();
+    }
+  }
+
+  function addWatchFiles(
+    this: Rollup.PluginContext,
+    fromId: string,
+    files: Set<string>,
+  ) {
+    // We don't need to watch files in build mode
+    if (config.command === 'build' && !config.build.watch) {
+      return;
+    }
+
+    const filesToWatch = [...files].filter(
+      (file) => !file.includes('node_modules'),
+    );
+
+    for (const file of filesToWatch) {
+      if (normalizePath(file) !== fromId) {
+        this.addWatchFile(file);
+      }
+    }
```

Comment:
Likely not performance-critical code, but you could just iterate over `files` once with a `for of` and put the `node_modules` check inside the loop, avoiding the creation of an array and the overhead from `filter`:

```suggestion
    for (const file of files) {
      if (!file.includes('node_modules') && normalizePath(file) !== fromId) {
        this.addWatchFile(file);
      }
    }
```

Reply:
Good suggestion 👍 This code went through a few stages of refactoring.